fastn-core/src/migrations.rs

```diff
@@ -0,0 +1,224 @@
+pub(super) const MIGRATION_TABLE: &str = r#"
+
+CREATE TABLE IF NOT EXISTS
+    fastn_migration
+(
+    id               INTEGER PRIMARY KEY,
+    app_name         TEXT NOT NULL,
+    migration_number INTEGER NOT NULL UNIQUE,
+    migration_name   TEXT NOT NULL,
+    applied_on       INTEGER NOT NULL
+) STRICT;
+
+"#;
+
+pub(crate) async fn create_migration_table(
+    config: &fastn_core::Config,
+    db: &str,
+) -> Result<(), fastn_utils::SqlError> {
+    config.ds.sql_batch(db, MIGRATION_TABLE).await?;
+    Ok(())
+}
+
+pub(crate) async fn migrate(
+    req_config: &mut fastn_core::RequestConfig,
+    db: &str,
+) -> Result<(), fastn_utils::SqlError> {
+    let migration_ftd = get_migrations_from_ftd(req_config).await;
+    let migration_sql = get_migration_from_sql(&req_config.config, db).await?;
+
+    let migrations_to_apply = find_migrations_to_apply(&migration_ftd, &migration_sql).await;
+
+    let now = chrono::Utc::now();
+    for migration in migrations_to_apply {
+        req_config
+            .config
+            .ds
+            .sql_batch(db, migration.content.as_str())
+            .await?;
+        mark_migration_applied(&req_config.config, db, &migration, &now)
+    }
+    Ok(())
+}
+
+async fn mark_migration_applied(
+    config: &fastn_core::Config,
+    db: &str,
+    migration_data: &MigrationDataFTD,
+    now: &chrono::DateTime<chrono::Utc>,
+) -> Result<(), fastn_utils::SqlError> {
+    let now_in_nanosecond = now.timestamp_nanos_opt().unwrap();
+    config
+        .ds
+        .sql_execute(
+            db,
+            format!(
+                r#"
+                    INSERT INTO
+                        fastn_migration
+                            (app_name, migration_number, migration_name, applied_on)
+                    VALUES
+                        ({}, {}, {}, {});
+                "#,
+                config.package.name, migration_data.number, migration_data.name, now_in_nanosecond
+            )
+            .as_str(),
+            vec![],
+        )
+        .await?;
+    Ok(())
+}
+
+struct MigrationDataFTD {
+    number: i64,
+    name: String,
+    content: String,
+}
+
+async fn get_migrations_from_ftd(
+    req_config: &mut fastn_core::RequestConfig,
+) -> Vec<MigrationDataFTD> {
+    let migration_ftd = req_config
+        .config
+        .package
+        .fs_fetch_by_file_name("migration.ftd", None, &req_config.config.ds)
+        .await?;
```

Comment:
Nopes, migraiton must be `FASTN.ftd` for now. If we allow this we have to open and parse one more file on every page view.